### ISSUE TITLE
Suggestion: Auto-load aircraft performance profile based on path

### ISSUE THREAD

**Issue by JT8D-17** [albar965/littlenavmap#1113]
It would be a minor quality of life improvement if LittleNavMap stored paths to aircraft in its aircraft performance profiles and automatically loaded them if the associated aircraft in was loaded in the simulator.

Here's a suggested design outline for such a feature for X-Plane (I'm sure there's similar methods in other simulators):

- Reading the path to the user aircraft can be done with [XPLMGetNthAircraftModel](https://developer.x-plane.com/sdk/XPLMGetNthAircraftModel/), with "inIndex" = 0. "outPath" returns the path to the ACF file.

- LittleXPConnect would read the path at every aircraft reload or flight start event and then either:
   - iterate through all ".lnmperf" files to find a matching path
or
   - access a database file that contains the path to an ACF file and its associated .lnmperf file (to avoid iterating through files).
  
- Upon finding a matching lnmperf file, said file would be loaded. No match naturally would of course not do anything.

- This would be done at LNM startup or aircraft reloads (changes) triggered in X-Plane.

- Adding a path would be done in the aircraft performance editor by clicking an "Add Path" button or filling in the text field manually.

- Supporting multiple paths per performance files would also be possible so that different ACF files for the same aircraft (e.g. passenger and cargo variants) can access the same data.

**Comment by albar965** [comment]
Already on the list.

The aircraft model (ICAO model) is already read from .acf or aircraft.cfg files as well as the model path. The latter one is not needed since the ICAO model is enough. 

I'd rather scan through a path defined by the user or all files in the history.
Also do this on demand and in background since it would delay the startup.

**Comment by albar965** [comment]
Nope. LNM did never auto-load the profiles.

Thinking of this, it does not make much sense IMO since normally one creates a plan before sitting in the cockpit at the departure. There is no ICAO aircraft type transferred to LNM before. 

Taking into account engine differences is not possible at all, as LNM cannot detect them in any simulator.

I think the better solution is to load it manually.

**Comment by JT8D-17** [comment]
> Thinking of this, it does not make much sense IMO since normally one creates a plan before sitting in the cockpit at the departure. There is no ICAO aircraft type transferred to LNM before.

Not necessarily. I mostly let the flight load at the departure airport and only then decide where to go.

> Taking into account engine differences is not possible at all, as LNM cannot detect them in any simulator.

In X-Plane, you most definitely can by reading the path down to the ACF file.
Here's why: Unless the aircraft uses some plugin magic to alter engine parameters (thrust, RPMs, fuel flow, etc.) to cover different engine variants, the only other way is to use different ACF files (e.g. 747-400_PW.acf, 747-400_RR.acf, etc.) with specific engine parameters, but which are stored in the exact same folder because they use (almost) the exact same art assets.
(FSX behaving exactly the same, IIRC.)

The alternative of altering the ICAO code to some fantasy one to represent a certain engine variant in order to autoload a suitable performance file in LNM would mess up other simulator elements like ATC.

**Comment by albar965** [comment]
acf does not contain engine variants and there is no way I'll read the filesnames (which is unreliable anway) to get the variants. Not worth the effort and added complexity.

**Comment by JT8D-17** [comment]
I'm using "[XPLMGetNthAircraftModel](https://developer.x-plane.com/sdk/XPLMGetNthAircraftModel/)" to get the path in a few projects of mine and it works absolutely reliably each and every time.

**Comment by albar965** [comment]
I already do the same since not all parameters are available as datarefs but I'm not going to scan the filename for engine types.

**Comment by JT8D-17** [comment]
> I already do the same since not all parameters are available as datarefs but I'm not going to scan the filename for engine types.

I think we have a grave misunderstanding here. "Engine variants" was just a use case to justify using the path to an ACF or aircraft.cfg file as a definite association for a performance profile.

**Comment by JT8D-17** [comment]
Just got an idea how to improve performance file loading user experience.

LNM compares the ICAO of the currently loaded performance file with trhe user aircraft's.
If it does not match, LNM opens a text box:
"Aircraft ICAO identifier mismatch detected  (old: xxx, current: yyy). Loaded performance data may deliver wrong results.
Caution: Loading or crating a new file will reset the currently collected performance data!

[Load Perf File] [Create New Perf File] [Cancel]"

Loading a new perforemance file will open the file selection dialog, creating a new one will open the corresponding window and cancelling will dismiss the box.

**Comment by albar965** [comment]
> Just got an idea how to improve performance file loading user experience.
> 
> LNM compares the ICAO of the currently loaded performance file with trhe user aircraft's. If it does not match, LNM opens a text box: "Aircraft ICAO identifier mismatch detected (old: xxx, current: yyy). Loaded performance data may deliver wrong results. Caution: Loading or crating a new file will reset the currently collected performance data!
> 
> [Load Perf File] [Create New Perf File] [Cancel]"
> 
> Loading a new perforemance file will open the file selection dialog, creating a new one will open the corresponding window and cancelling will dismiss the box.

Already there:

![Image](https://github.com/user-attachments/assets/641793b8-7cb4-4cba-9c41-e0d03b615eb7)

![Image](https://github.com/user-attachments/assets/8a4684ec-d57b-44e0-98f2-0e7f8fcf0774)

Alex

**Comment by JT8D-17** [comment]
Thanks, but that's absolutely not what I meant.

I think this is going nowhere. Thank you for the time.

**Comment by albar965** [comment]
> Thanks, but that's absolutely not what I meant.

Mind to clarify? I can avoid the reset of the collection at least.

**Comment by JT8D-17** [comment]
> Mind to clarify? I can avoid the reset of the collection at least.

As it is now, I find the aircraft performance handling in the user interface generally unengaging and clumsy. If I switch aircraft in-sim, I have to navigate to LNM's window, click the performance tab, click "open", scroll down to another performance file, open the file, then go to the data collection tab, check that there's no data left in there or reset it, then go back to actually planning a flight.
A warning message about a mismatch simply does not help here because I have to do the same clicks anyway.
It's a similar misery with collected performance data. Nothing reminds me to review it after a flight and I've lost or messed up a good chunk of collected data over the years because of that.

So ideally:
- LNM would detect a discrepancy between ICAO codes and say "You've switched aircraft from x to y. I found a profile for y." and I'd just click "Load". Without a found profile, I'd just click "Select" or "Create New". (Or when I'm not in the mood for any aircraft performance considerations at all I'd just click "No Thanks" and disable aircraft performance handling whatsoever.)
- Performance data collection happens within the selected profile and not outside of it, i.e. two sets of performance data: "accepted" (exsiting "perf" tags in the XML), used for flight plan calculation and "collected", storing session data. After a flight and shutdown (if detected), there would be a "I have collected performance data in this session. What should I do with it?" popup. "Review" would send me to something akin to the current dialogue where I can merge, copy, etc., "Discard" would throw away the collected data and "Nothing" would simply close the popup.

I know this would require extensive changes to the UI and core systems and an extra dataset in the profile and all. Maybe there's another way to improve this. All I wish is that LNM was...smarter when it comes to a/c performance.

**Comment by albar965** [comment]
> As it is now, I find the aircraft performance handling in the user interface generally unengaging and clumsy. If I switch aircraft in-sim, I have to navigate to LNM's window, click the performance tab, click "open", scroll down to another performance file, open the file, then go to the data collection tab, check that there's no data left in there or reset it, then go back to actually planning a flight.

You can simply select a perf file from the recent list in the main menu. No need to open the performance tab and click through file dialogs. Why make it as complicated as you describe it?
Also I don't understand why the data collection is needed all the time. Usually it is done after a few flights and can be ignored once a performance profile is refined. 

> * LNM would detect a discrepancy between ICAO codes and say "You've switched aircraft from x to y. I found a profile for y." and I'd just click "Load". Without a found profile, I'd just click "Select" or "Create New". (Or when I'm not in the mood for any aircraft performance considerations at all I'd just click "No Thanks" and disable aircraft performance handling whatsoever.)

I'll have a look at this. Maybe better than the initial proposed automatic loading. Simply scan folders after detecting a new aircraft and give the user a choice to load, select or create new.

Adding a question to reset the collection or not when changing aircraft or perf file can make sense, agree.

Everything covered? 🙂 

Alex

**Comment by JT8D-17** [comment]
> You can simply select a perf file from the recent list in the main menu. No need to open the performance tab and click through file dialogs. Why make it as complicated as you describe it? 

Because I've (knowingly) never checked out the "Aircraft" menu before and I've been using LNM for...when was your first release?
The menu item is a bit lost all the way to the right. It might have helped if the menu items were sorted alphabetically (save for "File" and "Help" because these are always first and last in menu bars) as I probably did not expect an entry starting with "A" being so far to the right.

> Also I don't understand why the data collection is needed all the time. Usually it is done after a few flights and can be ignored once a performance profile is refined.

Yes and no. One may have reasonable, general figures after a few flights, but as different cruise altitudes, speeds, fuel and payloads yield different performance, so there is really not a "one size fits all" solution. Then the sim platform might introduce changes that affect fuel burn, the aircraft author might refine (or mess) up fuel burn with an update, etc...so it's better to collect, review and if needed merge every flight.


> I'll have a look at this. Maybe better than the initial proposed automatic loading. Simply scan folders after detecting a new aircraft and give the user a choice to load, select or create new.
> 
> Adding a question to reset the collection or not when changing aircraft or perf file can make sense, agree.
> 
> Everything covered? 🙂

Yes, that'd be a major improvement and much appreciated.

**Comment by albar965** [comment]
Ok. But I see no sorted menu items in other programs. Would not make sense anyway. This would change the order in other languages.
